server/src/main/scala/org/http4s/server/staticcontent/WebjarService.scala

```diff
@@ -55,7 +55,9 @@ object WebjarService {
     * @param config The configuration for this service
     * @return The HttpRoutes
     */
-  def apply[F[_]](config: Config[F])(implicit F: Sync[F], cs: ContextShift[F]): HttpRoutes[F] = {
+  def apply[F[_]](config: Config[F], classLoader: Option[ClassLoader] = None)(
```

Comment:
Is there any reason to introduce another parameter rather than put this in `Config`?

I have a more radical idea: move this to the same kind of builder pattern we use for server builders, and unlike the case class config or default optional parameters, we'll be able to introduce more parameters without breaking compatibility in the future.

Reply:
If this doesn't make any sense, I'd be happy to sketch one, and could follow the pattern on the other.

Reply:
@rossabaker A builder like this??

```
  class WebjarServiceBuilder[F[_]] {
    private var webjarAssetFilter: WebjarAssetFilter = _ => true
    private var cacheStrategy: CacheStrategy[F] = NoopCacheStrategy[F]
    private var classLoader: Option[ClassLoader] = None

    def wid(webjarAssetFilter: WebjarAssetFilter): WebjarServiceBuilder[F] = {
      this.webjarAssetFilter = webjarAssetFilter
      this
    }

    def wid(cacheStrategy: CacheStrategy[F]): WebjarServiceBuilder[F] = {
      this.cacheStrategy = cacheStrategy
      this
    }

    def wid(classLoader: ClassLoader): WebjarServiceBuilder[F] = {
      this.classLoader = Some(classLoader)
      this
    }

    def build(implicit F: Sync[F], cs: ContextShift[F]): HttpRoutes[F] = ???
  }
```

Reply:
That's pretty close.  This is pretty similar to a builder in an object-oriented language, but our backend "builders" are immutable.  Something like:

```scala
// private constructor: we want people to call it like `WebjarServiceBuilder[F].blah.build`
class WebjarServiceBuilder[F[_]] private (
  webjarAssetFilter: WebjarAssetFilter,
  cacheStrategy: CacheStrategy[F],
  classLoader: Option[ClassLoader]) {

  // You don't strictly need this, but it means you don't need to repeat arguments in all your `with*` methods below. 
  private def copy(
    webjarAssetFilter: WebjarAssetFilter = webjarAssetFilter,
    cacheStrategy: CacheStrategy[F] = cacheStrategy,
    classLoader: Option[ClassLoader] = classLoader) =
    new WebjarServiceBuilder(webjarAssetFilter, cacheStrategy, classLoader)

  // Each field gets its own "with" setter.
  def withWebjarAssetFilter(webjarAssetFilter: WebjarAssetFilter) = 
    copy(webjarAssetFilter = webjarAssetFilter)

  // and so on for cacheStrategy and classLoader

  // Or maybe call this `.toRoutes`?
  def build = ???
}

object WebjarServiceBuilder {
  // If you have arguments with no sensible default, pass them to apply.
  // This is also a good spot for constraints, if you have any.
  def apply[F[_]] = new WebjarServiceBuilder(
     // Supply your default arguments here
     webjarAssetFilter = _ => true,
     cacheStrategy = NoopCacheStrategy[F],
     classLoader = None)
}
```

I haven't tried to compile any of that, but look at the other files named *Builder for working examples.

Reply:
@rossabaker cool, got it

Reply:
@rossabaker do you want to get rid of the class `WebjarService.Config[F[_}]` all together? That will impact the way it is interfaced with the outside world. [Here](https://github.com/http4s/http4s/blob/0ec8e6a69da2bd53774bdc633eb768b336ec7b8a/server/src/main/scala/org/http4s/server/staticcontent/package.scala#L23) is the line. The users of the library will now have to change their way of calling `wejarService[F](..)` method.